subsys/net/lib/zstream/zstream_tls.c

```diff
@@ -0,0 +1,228 @@
+	errno = EINVAL;
+	return -1;
+}
+
+static int zstream_tls_flush(zstream stream)
+{
+	/* Believe or not, but mbedTLS doesn't buffer outbut data,
+	 * even for a single byte written it will create a TLS record
+	 * and send at once.
+	 */
+	return 0;
+}
+
+int zstream_tls_close(zstream stream)
+{
+	struct zstream_tls *self = (struct zstream_tls *)stream;
+	int tls_res = mbedtls_ssl_close_notify(&self->ssl);
+	/* We need to close underlying stream regardless of TLS close
+	 * notify status, or the stream will be leaked.
+	 */
+	int stream_res = zstream_close(self->sock);
+
+	if (tls_res == 0 && stream_res == 0) {
+		return 0;
+	}
+
+	if (stream_res == 0) {
+		/* TODO: Check which mbedTLS errors are possible and improve
+		 * mapping to POSIX codes.
+		 */
+		errno = EINVAL;
+	}
+
+	return -1;
+}
+
+static const struct zstream_api zstream_tls_api = {
+	.read = zstream_tls_read,
+	.write = zstream_tls_write,
+	.flush = zstream_tls_flush,
+	.close = zstream_tls_close,
+};
+
+static int zstream_mbedtls_ssl_send(void *ctx, const unsigned char *buf,
+			     size_t len)
+{
+	zstream sock = ctx;
+	ssize_t outlen = zstream_write(sock, buf, len);
+
+	if (outlen != -1) {
+		return outlen;
+	}
+
+	if (errno == EAGAIN) {
+		return MBEDTLS_ERR_SSL_WANT_WRITE;
+	}
+
+	/* Generic error */
+	return MBEDTLS_ERR_NET_SEND_FAILED;
+}
+
+static int zstream_mbedtls_ssl_recv(void *ctx, unsigned char *buf,
+				    size_t len)
+{
+	zstream sock = ctx;
+	ssize_t outlen = zstream_read(sock, buf, len);
+
+	if (outlen != -1) {
+		return outlen;
+	}
+
+	if (errno == EAGAIN) {
+		return MBEDTLS_ERR_SSL_WANT_READ;
+	}
+
+	/* Generic error */
+	return MBEDTLS_ERR_NET_RECV_FAILED;
+}
+
+// TODO: FIXME!
+static int null_entropy_func(void *data, unsigned char *output, size_t len) {
+	(void)data;
+	(void)output;
+	(void)len;
+	// enjoy random bytes
+	return 0;
+}
+
+int zstream_tls_init(struct zstream_tls *p, zstream sock, bool server_side)
+{
+	int ret;
+	static const unsigned char seed[] = "zephyr";
+
+	p->api = &zstream_tls_api;
+	p->sock = sock;
+
+	mbedtls_ssl_init(&p->ssl);
+	mbedtls_x509_crt_init(&p->cacert);
+	mbedtls_x509_crt_init(&p->cert);
+	mbedtls_pk_init(&p->pkey);
+
+	/* Entropy, sec */
+	mbedtls_entropy_init(&p->entropy);
+	mbedtls_ctr_drbg_init(&p->ctr_drbg);
+	ret = mbedtls_ctr_drbg_seed(&p->ctr_drbg, null_entropy_func, &p->entropy, seed, sizeof(seed));
+	if (ret != 0) {
+		goto cleanup;
+	}
+
+	/* sec */
+	mbedtls_ssl_config_init(&p->conf);
+	ret = mbedtls_ssl_config_defaults(&p->conf,
+		server_side ? MBEDTLS_SSL_IS_SERVER : MBEDTLS_SSL_IS_CLIENT,
+		MBEDTLS_SSL_TRANSPORT_STREAM,
+		MBEDTLS_SSL_PRESET_DEFAULT);
+	if (ret != 0) {
+		goto cleanup;
+	}
+
+	mbedtls_ssl_conf_authmode(&p->conf, MBEDTLS_SSL_VERIFY_NONE); // FIXME
+	mbedtls_ssl_conf_rng(&p->conf, mbedtls_ctr_drbg_random, &p->ctr_drbg);
+
+	ret = mbedtls_ssl_setup(&p->ssl, &p->conf);
+	if (ret != 0) {
+		goto cleanup;
+	}
+
+//	ret = mbedtls_ssl_set_hostname(&o->ssl, sni);
```

Comment:
For cert-based authentication, this applies just to the client, and sets the name that the server's certs will be verified against. This makes sense if the server is a "real" network host, less so if it is something like a gateway, that won't have a meaningful hostname.